### ISSUE TITLE
Integrate in the Open MCT visualizer the battery status

### ISSUE THREAD

**Issue by nunoguedelha** [ami-iit/yarp-openmct#14]
@nunoguedelha commented on [Tue Jul 06 2021](https://github.com/dic-iit/element_software-engineering/issues/58)

---

@nunoguedelha commented on [Thu Aug 12 2021](https://github.com/dic-iit/element_software-engineering/issues/58#issuecomment-897810672)

Ciao @S-Dafarra , which ports do you read for getting the battery status? I didn't find anything in https://github.com/robotology/walking-teleoperation nor in https://github.com/robotology/walking-controllers. Could you provide me a link to the how-to if it exists eventually? Thx!

---

@S-Dafarra commented on [Thu Aug 12 2021](https://github.com/dic-iit/element_software-engineering/issues/58#issuecomment-897830082)

Ciao @nunoguedelha! For the battery we were launching a device. I added some details in https://github.com/robotology/icub-main/pull/727. Back in time, I was looking at the terminal output, but it would be possible to get the battery level also by looking at the output of the ``batteryWrapper`` (https://github.com/dic-iit/robots-configuration/blob/devel_iCubGenova04/iCubGenova04/battery/wrappers/battery/battery.xml#L6). Hence, I think this can be done in two ways, either by reading a port, or by streaming the output of the module (like with ``yarplogger``)

---

@nunoguedelha commented on [Thu Aug 12 2021](https://github.com/dic-iit/element_software-engineering/issues/58#issuecomment-897857955)

Probably reading the port should be easier, more straightforward. I guess there is no respective GazeboYarpPlugin device right?

---

@S-Dafarra commented on [Fri Aug 13 2021](https://github.com/dic-iit/element_software-engineering/issues/58#issuecomment-898214676)

> . I guess there is no respective GazeboYarpPlugin device right?

I don't think so

**Comment by traversaro** [comment]
You can use the `fakeBattery` device to simulate a battery even disconneted from Gazebo: http://www.yarp.it/git-master/classFakeBattery.html#details .

**Comment by nunoguedelha** [comment]
Great! And this should also be run through the single launch command script (https://github.com/dic-iit/yarp-openmct/issues/25).

**Comment by nunoguedelha** [comment]
## Testing 'yarpbatterygui' on the streaming port 'icub/battery'

The port that should stream the battery status data is `icub/battery`. Instead o running the battery service `battery_bluetooth.service` (https://github.com/robotology/icub-main/pull/727#issue-588122515), we run the fake battery, using the `fakeBattery` device to simulate a battery (disconneted from Gazebo) http://www.yarp.it/git-master/classFakeBattery.html#details, as suggested by @traversaro .

**Comment by nunoguedelha** [comment]
### Build "Fake battery" device

In the YARP cmake configuration, set to `ON` the following options:
```
ENABLE_yarpmod_fakeBattery
ENABLE_yarpmod_batteryWrapper
YARP_COMPILE_yarpbatterygui (default)
```

Now the command `yarpdev --list | grep Battery` returns
```
[INFO] |yarp.dev.Drivers| Device "fakeBattery", available on request (found in <path-to-robotology-superbuild>/build/install/lib/yarp/yarp_fakeBattery.so library), wrapped by "batteryWrapper".
```

### Use the "Fake batterie" and the Battery Wrapper devices in the initial **yarprobotinterface** startup configuration (along with wholeBodyDynamics configuration).

Find the `icub_battery.xml` configuration file
```
$ find <path-to-robotology-superbuild>/build/install -name icub_battery.xml
<path-to-robotology-superbuild>/build/install/share/ICUBcontrib/robots/iCubGenova04/battery/icub_battery.xml
```

Content...
```xml
<robot name="battery" build="1" portprefix="battery" xmlns:xi="http://www.w3.org/2001/XInclude">
<devices>
    <!-- battery -->
    <xi:include href="./wrappers/battery/battery.xml" />
    <xi:include href="./hardware/battery/icubbattery.xml" />

</devices>
</robot>
```

Wrapper configuration in the same folder in file `wrappers/battery/battery.xml`:
- wrapper name => "icubBattery"
- port name => "/icub/battery"
- device driver name => "icub_battery"

Devide driver configuration is in the same folder in file `hardware/battery/icubbattery.xml`.

**Comment by nunoguedelha** [comment]
- Renamed `icubbattery.xml` to `icubfakebattery.xml`.
- Changed device type to `fakeBattery` in `icubfakebattery.xml`.
- left unchanged all other parameters (they're probably not used. to be checked).
- Ran the `yarprobotinterface`
    ```
    YARP_CLOCK=/clock yarprobotinterface --config battery/icub_battery.xml
    ```
    which returned
    ```
    [INFO] |yarp.os.Port| Port /iiticublap199.home/yarprobotinterfa/92031/clock:i active at tcp://192.168.1.18:10135/
    [INFO] |yarp.os.Network| Success: port-to-port persistent connection added.
    ...
    [DEBUG] yarprobotinterface running happily
    ```
- Ran the `yarpbatterygui`
    ```
    yarpbatterygui --remote /icub/battery
    ```
    which returned
    ```
    [INFO] refresh_period parameter missing. Using default value: 10 s
    [INFO] Using local port: /batteryMonitor0:i
    [DEBUG] |yarp.device.batteryClient| (device batteryClient) (local "/batteryMonitor0:i") (remote "/icub/battery")
    [INFO] |yarp.os.Port| Port /batteryMonitor0:i/data:i active at tcp://192.168.1.18:10132/
    [INFO] |yarp.os.Port| Port /batteryMonitor0:i/rpc:o active at tcp://192.168.1.18:10133/
    [INFO] |yarp.os.impl.PortCoreInputUnit| Receiving input from /icub/battery/data:o to /batteryMonitor0:i/data:i using udp
    [INFO] |yarp.os.impl.PortCoreOutputUnit| Sending output from /batteryMonitor0:i/rpc:o to /icub/battery/rpc:i using tcp
    [INFO] |yarp.dev.PolyDriver| Created device <batteryClient>. See C++ class BatteryClient for documentation.
    ```

We get broadcast messages on the console about the battery level, but no other type of output. The window `yarpbatterygui` window stays minimized with no content. @S-Dafarra is this normal?
```
Broadcast Message from nunoguedelha@iiticublap199.home
        (/dev/ttys018) at 6:22 CEST...

WARNING: battery charge below 15%


Broadcast Message from nunoguedelha@iiticublap199.home
        (/dev/ttys018) at 6:23 CEST...

WARNING: battery charge below 12%


Broadcast Message from nunoguedelha@iiticublap199.home
        (/dev/ttys018) at 6:24 CEST...

WARNING: battery charge below 10%
```

Anyway, reading directly the port outputs apparently the correct data:
```
yarp read ... /icub/battery/data:o
```
returns
```
[INFO] |yarp.os.Port| Port /tmp/port/1 active at tcp://192.168.1.18:10142/
[INFO] |yarp.os.impl.PortCoreInputUnit| Receiving input from /icub/battery/data:o to /tmp/port/1 using tcp
30.0 3.0 5.17500000002588667058 20.0 5
30.0 3.0 5.17000000002588500081 20.0 5
30.0 3.0 5.16500000002588333103 20.0 5
30.0 3.0 5.16000000002588166126 20.0 5
30.0 3.0 5.15600000002588032544 20.0 5
30.0 3.0 5.15100000002587865566 20.0 5
```

So I assume the fake battery device and respective wrapper (same wrapper as for the real battery) are working properly. We can proceed.

**Comment by S-Dafarra** [comment]
> We get broadcast messages on the console about the battery level, but no other type of output. 

Those messages are generated by the ``BatteryWrapper`` device: https://github.com/robotology/yarp/blob/5baf0506fbb59908b44cd8644d4f04da532be973/src/devices/batteryWrapper/BatteryWrapper.cpp#L420-L431. This device communicates with the ``fakeBattery``. Apparently there is a logic where it simulates the battery discharge https://github.com/robotology/yarp/blob/5baf0506fb/src/devices/fakeBattery/fakeBattery.cpp#L81-L86. By default, it considers a current that will discharge the battery (https://github.com/robotology/yarp/blob/5baf0506fb/src/devices/fakeBattery/fakeBattery.cpp#L24), so those message are intentional.

> The window `yarpbatterygui` window stays minimized with no content. @S-Dafarra is this normal?

This is weird I guess. It might be a bug on MacOS, I would suggest opening an issue for that :thinking:

**Comment by nunoguedelha** [comment]
Yes, actually I was worried only about the empty `yarpbatterygui` window. I will open an issue as you suggested. Thanks!

**Comment by nunoguedelha** [comment]
### Adding the source configuration files for the fake battery

The configuration files for the battery device are installed to the `ICUBcontrib` folder related to the selected robot name<sup> [1](#f1) </sup>, from the source files in the `robots-configuration` repository folder. There are no configuration files in that repo for iCubGazeboV2_5. The configuration files for Gazebo are installed into shared iCub folders<sup> [2](#f2) </sup>. The configuration files for specific modules like wholeBodyDynamics estimators are installed to shared Yarp folders<sup> [3](#f3) </sup> and from specific feature repositories like `whole-body-estimators`.

If there was a folder dedicated to the model iCubGazeboV2_5 in `robots-configuration`, we could just add the battery config files there. Since it's not the case, it seems to e that the immediate more reasonnable approach would be to add those files in the `yarp-openmct` repository and install them in `<path-to-robotology-superbuild>/build/install/share/yarp/robots`. It should be enough for now, we could always revise the approach later.

<b id="f1">1. </b> `<path-to-robotology-superbuild>/build/install/share/ICUBcontrib/robots/<robot-name>/battery`
<b id="f2">2. </b> `<path-to-robotology-superbuild>/build/install/share/iCub/conf and .../share/iCub/robots/iCubGazeboV2_5`
<b id="f3">3. </b> `<path-to-robotology-superbuild>/build/install/share/yarp/robots/<robot-name>`

**Comment by S-Dafarra** [comment]
You can also add a local folder of yours to the yarp data dirs. I would avoid installing them with this repo, unless for specific testing purposes.

**Comment by nunoguedelha** [comment]
> I would avoid installing them with this repo, unless for specific testing purposes.

Why? I would do exactly the opposite, adding the local folder to YARP_DATA_DIRS only for testing or fast temporary implementation, even by setting the variable directly in the command for running the **yarprobotinterface**. (maybe I misunderstood and that's what you meant.)

The goal here is to have the capability to run the visualization tool while simulating the robot on Gazebo. And for that we have installation folders `.../build/install/share/ICUBcontrib/robots/iCubGazeboV2_5`, `.../build/install/share/yarp/robots/iCubGazeboV2_5` and so on no?

> You can also add a local folder of yours to the yarp data dirs.

I wanted to avoid that, this adds a step in the installation process, which messes with the environment of the user and is less seamless than installing configuration files for the iCub Gazebo model. Of course, as I mentioned earlier, I could add the setting of the `YARP_DATA_DIRS` on the fly in the **yarprobotinterface** command line, the same way we do with `YARP_CLOCK`, but it would be more messy.

**All in all, installing those config files seems to me a better approach for the following reasons:**
- The files installed in the `iCubGazeboV2_5` folder have no impact, unless you launch explicitly the robot interface with those config files and the `YARP_ROBOT_NAME = iCubGazeboV2_5`.
- Those config files can easily be found and removed, and the result is valid in all your ongoing terminal sessions, unlike the scenario where you use environment variables.
- In the worst case, if getting rid of the config files required reinstalling the superbuild, this should be safer than playing around with an ever growing set of environment variables.

**Comment by S-Dafarra** [comment]
Without getting too much into bikeshedding arguments, if you install them with this repo, you still need to make sure that the installation prefix is in the ``YARP_DATA_DIRS`` (from the previous comments, I understood you are using the same installation prefix of the superbuild, but this cannot be enforced). My point was that for the end-user, these files are going to be useless, plus they can generate some confusion.

Imagine this repo will be added to the superbuild, as I hope. Those files will be installed. Now, imagine I want to edit them because something is not working anymore. I am never going to guess that those files are actually generated by this repo.

Hence, to reformulate my point, I would avoid installing by default any file that is not necessary for the correct functioning of the tool.

Still, there might be cases, also for the end user, in which you want to run some tests. And these files can be very handy, then. But for this specific case, it would be better to have a dedicated CMake option, something like ``INSTALL_FAKE_DEVICE_FILES``. Even better, could be to have them all in a single directory that is not installed. The first search path of the ``ResourceFinder`` is ``pwd``. Hence, it should be just sufficient to run all the tests and devices from the testing folder.

**Comment by nunoguedelha** [comment]
Sorry for the long answer @S-Dafarra 😄 , if you don't have time to read everything, you can always skip to the recap section at the bottom.

> you still need to make sure that the installation prefix is in the `YARP_DATA_DIRS`

Absolutely. This shall be documented and shall be configured through cmake options.

> My point was that for the end-user, these files are going to be useless

Hum, I have to disagree on that. Any user might want to run the tool in simulation and see the full tool plots in action, even in simulation, not to specifically test/verify that the tool works, but just to run a demo even without having access to the robot. But we agree on the fact the user should be able to activate/deactivate the telemetry signals he wishes to vizualize.

> Now, imagine I want to edit them because something is not working anymore. I am never going to guess that those files are actually generated by this repo

Well, in a long term I think that the fake battery could be a default configuration of the `iCubGazeboV2_5` model, either having the device independently run by the **yarprobotinterface** process either run by the Gazebo process through a plugin. In such case, whatever the framework you are running which uses the fake battery device and configuration, we would know where they are defined, but this is another discussion...

In any case, the plan was to have a "simulation" pane in the control console (along with other GUI improvements to come) where the user can activate any process required for simulation, like the fake battery. So it should be easy to make it clear to the user that the battery signal being visualized is fake<sup> [(*)](#f*) </sup>.

That said, either you are **not** an advanced developer, and you submit an issue if you encounter problem, either you are an advanced developer trying to analyze the problem by yourself, and in that case you know how to find the source files, or might just try changing the installed ones as a first step in solving the issue. Right here, using an environment variable wouldn't make things easier for any of those users, specially if you consider that the `YARP_DATA_DIRS` is an ordered list and that order determines which files are actually used.

> Hence, to reformulate my point, I would avoid installing by default any file that is not necessary for the correct functioning of the tool.
> 
> Still, there might be cases, also for the end user, in which you want to run some tests. And these files can be very handy, then. But for this specific case, it would be better to have a dedicated CMake option, something like `INSTALL_FAKE_DEVICE_FILES`. Even better, could be to have them all in a single directory that is not installed. The first search path of the `ResourceFinder` is `pwd`. Hence, it should be just sufficient to run all the tests and devices from the testing folder.

As I mentioned earlier, these files are required if the user wants to visualize the full widget set, not only for debug purposes.
=> It makes sense to have a CMake option `INSTALL_FAKE_DEVICE_FILES`. I would prefer to leave it to ON by default, as it would require less a step in the installation procedure if the user wants to run everything. **But I'm ok to have it OFF by default** and specifying in the installation procedure that it should be turned ON if you want to visualize the battery status.
=> The files are in a single directory already. Indeed, for testing purposes only, the user/tester can choose to not install them and run the battery device from that path at his convenience.

<b id="f*"> (*) </b> By the way that can evolve in the future, as a proper battery simulation could be integrated in the Gazebo plugins, along with current control simulation (if we relate to recent work done in the lab for current control).

### Recap (proposal) on the requirements
In a normal installation and usage of the tool:
1. If the user wishes to visualize the battery status, he has to activate a CMake option (for instance `INSTALL_FAKE_DEVICE_FILES`).
2. If the CMake option is ON, the fake battery files are installed.
3. The installation path shall be set through the CMake options.
4. If the user chooses to install the dependencies through the **robotology-superbuild**,  he shall specify the superbuild "install" path.
5. If the user installs the dependencies manually (without the superbuild) and chooses a custom installation folder for the visualization tool, such folder should be added to the `YARP_DATA_DIRS`.

**These requirements shall be clearly documented in the README.**

In a near future (as we both hope), if the visualization tool repository is added to the superbuild as a sub-project, the last two requirements 4 and 5 should be discarded.

**Comment by S-Dafarra** [comment]
> ### Recap (proposal) on the requirements
> In a normal installation and usage of the tool:
> 
> 1. If the user wishes to visualize the battery status, he has to activate a CMake option (for instance `INSTALL_FAKE_DEVICE_FILES`).
> 2. If the CMake option is ON, the fake battery files are installed.
> 3. The installation path shall be set through the CMake options.
> 4. If the user chooses to install the dependencies through the **robotology-superbuild**,  he shall specify the superbuild "install" path.
> 5. If the user installs the dependencies manually (without the superbuild) and chooses a custom installation folder for the visualization tool, such folder should be added to the `YARP_DATA_DIRS`.
> 
> **These requirements shall be clearly documented in the README.**
> 
> In a near future (as we both hope), if the visualization tool repository is added to the superbuild as a sub-project, the last two requirements 4 and 5 should be discarded.

Sounds good!

**Comment by nunoguedelha** [comment]
I forgot to mention, there will be an "uninstall" target also.

**Comment by nunoguedelha** [comment]
## Add the battery status to the telemetry server and visualizer

- Create a reader port `/yarpjs/battery/data:i` and connect it to the port `/icubSim/battery/data:o`.
- The telemetry server reads the data from `/yarpjs/battery/data:i` and generates the telemetry data from it.
- Add battery status entry to the OpenMCT visualizer (update the telemetry JSON dictonary).

=> implemented in #30.

As I checked on the web client (browser) the logged telemetry server replies to the visualizer requests, I realized there was a problem with the JSON fields:
<img width="550" alt="image" src="https://user-images.githubusercontent.com/6848872/133168505-996e470a-9a4f-43f7-9427-3f91bf8e4e6d.png">

Fixing...

**Comment by nunoguedelha** [comment]
There was a minor bug (break missing). The telemetry entry i functional now:

<img width="550" alt="image" src="https://user-images.githubusercontent.com/6848872/133176465-6060364b-aa1a-4076-b251-83158821152a.png">

<img width="1124" alt="image" src="https://user-images.githubusercontent.com/6848872/133176411-7123dadf-3c74-48ef-acc7-d05797b71bfd.png">
